File: matlab/wbgt.m
%
% wbgt
%
%     class to retrieves WBGT data from the KNMI open data platfon
%
% modifications
%   16-jul-2026   JM    initial version
%   28-jul-2026   JM    locations added, 2nd yaxis added

classdef wbgt < handle

  properties (Constant)
    TIMEZONE = 'Europe/Amsterdam';
    APIKEY = 'eyJvcmciOiI1ZTU1NGUxOTI3NGE5NjAwMDEyYTNlYjEiLCJpZCI6IjBhZTg4MTI5MTAzOTQyN2FiMDVhNjM2NmNiOWQxMTE3IiwiaCI6Im11cm11cjEyOCJ9';
  end

  properties (Access = private)

    locationCode = [];
    locationName = [];

    startTime = [];
    endTime = [];
    
    api = [];
    wbgtData = [];

  end

  methods (Access = public)
    
    % constructor
    %
    %   set the default location and time

    function obj = wbgt(location)

      if nargin < 1, location = []; end
      if isempty(location), location = 'Gilze-Rijen'; end

      obj.setLocation(location);
      
      obj.startTime = datetime("now",'TimeZone',obj.TIMEZONE) - hours(6);
      obj.endTime = datetime("now",'TimeZone',obj.TIMEZONE);

      obj.api = OpenAPIData('wet_bulb_globe_temperature','3.0',obj.APIKEY);

    end

    % input
    %
    %     ask for the start and end - time

    function obj = input(obj)
  
      inputStr = sprintf('begin of WBGT data (%s) > ', string(obj.startTime, 'dd-MMM-yyyy HH:mm'));
      answer = input(inputStr','s');
      if ~isempty(answer),  obj.startTime = datetime(answer, 'InputFormat', 'dd-MMM-yyyy HH:mm',TimeZone=obj.TIMEZONE); end

      inputStr = sprintf('end of WBGT data   (%s) > ', string(obj.endTime, 'dd-MMM-yyyy HH:mm')');
      answer = input(inputStr,'s');
      if ~isempty(answer),  obj.endTime = datetime(answer, 'InputFormat', 'dd-MMM-yyyy HH:mm',TimeZone=obj.TIMEZONE); end

    end

    % setLocation
    %
    %     sets the location where the WBGT values should be obtained

    function obj = setLocation(obj,location)

      if strcmpi(location,"Voorschoten")
        obj.locationCode = 6215;
        obj.locationName = location;
      end

      if strcmpi(location,"De Kooy Airport")
        obj.locationCode = 6235;
        obj.locationName = location;
      end

      if strcmpi(location,"Schiphol Airport")
        obj.locationCode = 6240;
        obj.locationName = location;
      end

      if strcmpi(location,"Hoorn Terschellingt")
        obj.locationCode = 6251;
        obj.locationName = location;
      end

      if strcmpi(location,"De Bilt")
        obj.locationCode = 6260;
        obj.locationName = location;
      end

      if strcmpi(location,"Lelystad Airport")
        obj.locationCode = 6269;
        obj.locationName = location;
      end

      if strcmpi(location,"Leeuwarden Airport")
        obj.locationCode = 6270;
        obj.locationName = location;
      end

      if strcmpi(location,"Deelen Airport")
        obj.locationCode = 6275;
        obj.locationName = location;
      end

      if strcmpi(location,"Hoogeveen")
        obj.locationCode = 6279;
        obj.locationName = location;
      end

      if strcmpi(location,"Groningen Airport Eelde")
        obj.locationCode = 6280;
        obj.locationName = location;
      end

      if strcmpi(location,"Twenthe Airport")
        obj.locationCode = 6290;
        obj.locationName = location;
      end

      if strcmpi(location,"Vlissingen")
        obj.locationCode = 6310;
        obj.locationName = location;
      end

      if strcmpi(location,"Westdorpe")
        obj.locationCode = 6319;
        obj.locationName = location;
      end

      if strcmpi(location,"Wilhelminadorp")
        obj.locationCode = 6323;
        obj.locationName = location;
      end

      if strcmpi(location,"Hoek van Holland")
        obj.locationCode = 6330;
        obj.locationName = location;
      end

      if strcmpi(location,"Rotterdam Airport")
        obj.locationCode = 6344;
        obj.locationName = location;
      end

      if strcmpi(location,"Cabauw")
        obj.locationCode = 6348;
        obj.locationName = location;
      end

      if strcmpi(location,"Gilze-Rijen Airport")
        obj.locationCode = 6350;
        obj.locationName = location;
      end

      if strcmpi(location,"Herwijnen")
        obj.locationCode = 6356;
        obj.locationName = location;
      end

      if strcmpi(location,"Eindhoven Airport")
        obj.locationCode = 6370;
        obj.locationName = location;
      end

      if strcmpi(location,"Volkel")
        obj.locationCode = 6375;
        obj.locationName = location;
      end

      if strcmpi(location,"Maastricht Airport")
        obj.locationCode = 6380;
        obj.locationName = location;
      end

    end

    % readData
    %
    %     first find the files that are needed and next read the
    %     information into a table

    function obj = readData(obj)

      files = obj.api.list_files(obj.startTime,obj.endTime);
      nrItems = size(files,1);
      fprintf("%d files received ...\n",nrItems);

      obj.wbgtData = table('Size',[nrItems 3],'VariableTypes', {'datetime','double','double'},'VariableNames', {'time','wbgt','heatIndex'});
      obj.wbgtData.time.TimeZone = obj.TIMEZONE;

      modFactor = 1;
      if nrItems > 10, modFactor = 10; end
      if nrItems > 100, modFactor = 50; end
      
      % and for every file 

      for i=1:nrItems

        [obj.wbgtData.time(i),obj.wbgtData.wbgt(i),obj.wbgtData.heatIndex(i)] = obj.api.getWBGTdata(files(i).filename,obj.locationCode); 
        if (mod(i,modFactor) == 0), fprintf(1,"%d (of %d) files processed ...\n",i,nrItems); end
        if (i==nrItems), fprintf(1,"%d (of %d) files processed ...\n",i,nrItems); end

      end

      obj.wbgtData = obj.wbgtData(~isnat(obj.wbgtData.time),:);
      obj.wbgtData = sortrows(obj.wbgtData,'time');

    end

    % plot
    %
    %   makes a graph from the obtained wbgtData. With plotAerea, shaded
    %   areas can be set

    function obj = plot(obj,plotArea)
     
      if (nargin < 1), plotArea = []; end
      if isempty(plotArea), plotArea = true; end

      figure
      theme(gcf,"light")

      if (size(obj.wbgtData,1) < 50)
        hold on
        plot(obj.wbgtData.time,obj.wbgtData.wbgt,'*','MarkerSize',8,'Color',[0 0 1]);
        plot(obj.wbgtData.time,obj.wbgtData.wbgt,'LineWidth',1,'Color',[0 0 1]);
        hold off
      else
        plot(obj.wbgtData.time,obj.wbgtData.wbgt,'LineWidth',1.5,'Color',[0 0 1]);  
      end

      set(gca,'Box','on')

      xlabel('Time',FontWeight='Bold');
      
      yyaxis left
      ylim([16 34]);
      ylabel('WBGT (^oC)',FontWeight='Bold');
      yticks([16 18 20 22 24 26 28 30 32 34]);

      yyaxis right
      ylim([16 34])
      ax = gca;
      ax.YColor = [0 0 0];

      ylabel('hittekracht',FontWeight='Bold',Color=[0 0 0]);
      yticks([16 18 20 22 24 26 28 30 32]);
      yticklabels({'2','3','4','5','6','7','8','9','10'})

      title(sprintf("WBGT at %s",obj.locationName),FontWeight='Bold')

      % plot the shaded areas, See slide Hein Daanen

      if plotArea

        hold on

        yregion(32.15,34,'FaceColor',[1 0 0],'FaceAlpha',0.7);
        yregion(30.05,32.15,'FaceColor',[1 0 0],'FaceAlpha',0.5);
        yregion(27.85,30.05,'FaceColor',[1 0 0],'FaceAlpha',0.3);
        yregion(25.65,27.85,'FaceColor',[0.95 0.6 0],'FaceAlpha',0.45);
        yregion(22.25,25.65,'FaceColor',[0.9 0.9 0],'FaceAlpha',0.25);
        yregion(18.4,22.25,'FaceColor',[0.9 0.9 0],'FaceAlpha',0.15);

        yline(32.15,'k-','LineWidth',0.5);
        yline(30.05,'k-','LineWidth',0.5);
        yline(27.85,'k-','LineWidth',0.5);
        yline(25.65,'k-','LineWidth',0.5);
        yline(22.25,'k-','LineWidth',0.5);
        yline(18.4,'k-','LineWidth',0.5);

        hold off
      end
    end

    % list
    %
    %   makes a list of the WBGT data

    function obj = list(obj)

      disp(obj.wbgtData);
    
    end

  end
end

File: python/wbgt.py
# WBGT
#
#		reads, list and plot the WBGT data at a certain weather station
#
# mdofications
#   16-jul-2026  JM  initial version 

#-jm from datetime import datetime, timedelta, timezone
#-jm import matplotlib.pyplot as plt
#-jm import matplotlib.dates as mdates
#-jm from python.opendata import OpenDataAPI

class wbgt :

	# init
	#
	#		initialises the class
	#			- sets the default location
	#			- sets the API key
	#		  - set default start and stop dates

	#-jm def __init__(self, location = "Gilze-Rijen") :

		# set the API key and the location

	#-jm	self.apiKey = "eyJvcmciOiI1ZTU1NGUxOTI3NGE5NjAwMDEyYTNlYjEiLCJpZCI6IjBhZTg4MTI5MTAzOTQyN2FiMDVhNjM2NmNiOWQxMTE3IiwiaCI6Im11cm11cjEyOCJ9"
	#-jm	self.setLocation(location)

		# use the OpenAPIdata class

	#-jm	self.api = OpenDataAPI(dataset = "wet_bulb_globe_temperature", version = "3.0", token = self.apiKey)

		# sets the default start and stop date
		
	#-jm	tz = timezone(timedelta(hours=2))
	#-jm	self.startTime = datetime.now(tz) - timedelta(hours = 6)
	#-jm	self.endTime = datetime.now(tz)

	# input
	#
	#    asks the user for start and stop dates

	#-jm def input(self) :

		#-jm inputString = "begin of WBGT data : (" + self.startTime.strftime("%d-%b-%Y %H:%M") + ") > "
		#-jmanswer = input(inputString)
		#-jmif answer != "" :
		#-jm		self.startTime = datetime.strptime(answer, "%d-%b-%Y %H:%M")

		#-jminputString = "end of WBGT data : (" + self.endTime.strftime("%d-%b-%Y %H:%M") + ") > "
		#-jmanswer = input(inputString)
		#-jmif answer != "" :
		#-jm	self.endTime = datetime.strptime(answer, "%d-%b-%Y %H:%M")

	# setLocation
	#
	#		sets the location for which the data should be retrieved	

	#-jm def setLocation(self,location) :

		#-jm if location.upper() == "GILZE-RIJEN" :
		#-jm	self.locationCode = "06350"
		#-jm	self.locationName = location

	# readData
	#
	#		reads tthe data from the KNMI sites
	#			- first read a list of CSV files that should be processed
	#			- process each file
	#		at the end the wbgtTable is filled

	#-jm def readData(self) :

		#-jm files = self.api.list_files(self.startTime, self.endTime)
		#-jm  nrItems = len(files);
		#-jm print("%d files received ..." % nrItems)

		#-jm modFactor = 1
		#-jm if nrItems > 10 :
		#-jm 	modFactor = 10
		#-jm if nrItems > 100 :
		#-jm 	modFactor = 50

		#-jm self.wbgtTable = []
		#-jm for item,i in zip(files,range(1,nrItems+1)) :

			#-jm [time,wbgt,heatIndx] = self.api.getWBGTdata(item["filename"],self.locationCode)
			#-jm self.wbgtTable.append({"time" : time, "wbgt" : wbgt, "heatIndex" : heatIndx})

			# show process (not for all files)
	
			#-jm if ((i % modFactor) == 0) or (i==nrItems) :
			#0jm 	print("%d (of %d) files processed ..." % (i,nrItems))

		#-jm self.wbgtTable = sorted(self.wbgtTable, key=lambda x: x["time"])
		
	# list
	#
	#		lists the data in the wbgtTable		

	#-jm def list(self) :

		#-jm print(f"{'Time':<20} {'WBGT':>8} {'HeatIndex':>10}")
		#-jm print("-" * 40)

		#-jm for row in self.wbgtTable :
		#-jm 	print(
		#-jm		f"{row['time'].strftime('%d-%m-%Y %H:%M'):<20} "
		#-jm		f"{row['wbgt']:>8.1f} "
		#-jm 		f"{row['heatIndex']:>5.0f}"
		#-jm	)
	
	# plot
	#
	#		plots the data in the wbgtTable

	#-jm def plot (self,pltAreas = True) :
		
		# get the x and y data 
		#-jm time = [row["time"] for row in self.wbgtTable]
		#-jm wbgt = [row["wbgt"] for row in self.wbgtTable]

		#-jm fig, ax = plt.subplots()
		#-jm fig.set_facecolor("white")

		#-jm if len(time) < 50 :
		#=jm	ax.plot(time,wbgt,"*",markersize=6,color="blue")
		#-jm 	ax.plot(time,wbgt,linewidth=1,color="blue")
		#-jm else :
		#-jm 	ax.plot(time,wbgt,linewidth=1.5,color="blue")

		#-jm ax.xaxis.set_major_formatter(mdates.DateFormatter('%H:%M'))
		#-jm ax.xaxis.set_major_locator(mdates.HourLocator(interval=1))
		#-jm ax.xaxis.set_major_formatter(mdates.DateFormatter('%H:%M'))

		#-jm ax.set_ylim(16, 34)
		#-jm ax.set_xlabel("Time", fontweight="bold")
		#-jm ax.set_ylabel("WBGT (°C)", fontweight="bold")
		#-jm ax.set_title("WBGT at " + self.locationName + " (" + time[-1].strftime("%d-%b-%Y") +")",fontweight="bold")

		# box around plot

		#-jm for spine in ax.spines.values() :
			#-jm spine.set_visible(True)	

  	# if plotAreas is set shaded areas are plotted
	
		#-jm if pltAreas :

			# shaded areas

			#-jmax.axhspan(32.15, 34, facecolor="red", alpha=0.7)
			#-jm ax.axhspan(30.05, 32.15, facecolor="red", alpha=0.5)
			#-jm ax.axhspan(27.85, 30.05, facecolor="red", alpha=0.3)
			#-jm ax.axhspan(25.65, 27.85, facecolor=(0.95, 0.6, 0), alpha=0.45)
			#-jm ax.axhspan(22.25, 25.65, facecolor=(0.9, 0.9, 0), alpha=0.25)
			#-jm ax.axhspan(18.4, 22.25, facecolor=(0.9, 0.9, 0), alpha=0.15)

			# border lines
		
			#-jm ax.axhline(32.15, color='k', lw=0.5)	
			#-jm ax.axhline(30.03, color='k', lw=0.5)	
			#-jm ax.axhline(27.85, color='k', lw=0.5)	
			#-jm ax.axhline(25.65, color='k', lw=0.5)	
			#-jm ax.axhline(22.25, color='k', lw=0.5)	
			#-jm ax.axhline(18.4, color='k', lw=0.5)	
		
		# and show

		#-jm plt.show()

